// @xmtp/xmtp-js/dist/cjs/src/crypto/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Signature = exports.PrivateKeyBundle = exports.PrivateKey = exports.PublicKeyBundle = exports.PublicKey = exports.decrypt = exports.encrypt = exports.utils = void 0;
const PublicKeyBundle_1 = __importDefault(require("./PublicKeyBundle"));
exports.PublicKeyBundle = PublicKeyBundle_1.default;
const PrivateKey_1 = __importDefault(require("./PrivateKey"));
exports.PrivateKey = PrivateKey_1.default;
const PrivateKeyBundle_1 = __importDefault(require("./PrivateKeyBundle"));
exports.PrivateKeyBundle = PrivateKeyBundle_1.default;
const PublicKey_1 = __importDefault(require("./PublicKey"));
exports.PublicKey = PublicKey_1.default;
const Signature_1 = __importDefault(require("./Signature"));
exports.Signature = Signature_1.default;
const utils = __importStar(require("./utils"));
exports.utils = utils;
const encryption_1 = require("./encryption");
Object.defineProperty(exports, "encrypt", { enumerable: true, get: function () { return encryption_1.encrypt; } });
Object.defineProperty(exports, "decrypt", { enumerable: true, get: function () { return encryption_1.decrypt; } });
