// js-waku/build/esm/proto/waku/v2/filter.js
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";
import { WakuMessage } from "../../waku/v2/message";
export const protobufPackage = "waku.v2";
function createBaseFilterRequest() {
    return { subscribe: false, topic: "", contentFilters: [] };
}
export const FilterRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.subscribe === true) {
            writer.uint32(8).bool(message.subscribe);
        }
        if (message.topic !== "") {
            writer.uint32(18).string(message.topic);
        }
        for (const v of message.contentFilters) {
            FilterRequest_ContentFilter.encode(v, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFilterRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.subscribe = reader.bool();
                    break;
                case 2:
                    message.topic = reader.string();
                    break;
                case 3:
                    message.contentFilters.push(FilterRequest_ContentFilter.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            subscribe: isSet(object.subscribe) ? Boolean(object.subscribe) : false,
            topic: isSet(object.topic) ? String(object.topic) : "",
            contentFilters: Array.isArray(object === null || object === void 0 ? void 0 : object.contentFilters)
                ? object.contentFilters.map((e) => FilterRequest_ContentFilter.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.subscribe !== undefined && (obj.subscribe = message.subscribe);
        message.topic !== undefined && (obj.topic = message.topic);
        if (message.contentFilters) {
            obj.contentFilters = message.contentFilters.map((e) => e ? FilterRequest_ContentFilter.toJSON(e) : undefined);
        }
        else {
            obj.contentFilters = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseFilterRequest();
        message.subscribe = (_a = object.subscribe) !== null && _a !== void 0 ? _a : false;
        message.topic = (_b = object.topic) !== null && _b !== void 0 ? _b : "";
        message.contentFilters =
            ((_c = object.contentFilters) === null || _c === void 0 ? void 0 : _c.map((e) => FilterRequest_ContentFilter.fromPartial(e))) || [];
        return message;
    },
};
function createBaseFilterRequest_ContentFilter() {
    return { contentTopic: "" };
}
export const FilterRequest_ContentFilter = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.contentTopic !== "") {
            writer.uint32(10).string(message.contentTopic);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFilterRequest_ContentFilter();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.contentTopic = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            contentTopic: isSet(object.contentTopic)
                ? String(object.contentTopic)
                : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.contentTopic !== undefined &&
            (obj.contentTopic = message.contentTopic);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseFilterRequest_ContentFilter();
        message.contentTopic = (_a = object.contentTopic) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseMessagePush() {
    return { messages: [] };
}
export const MessagePush = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.messages) {
            WakuMessage.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMessagePush();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.messages.push(WakuMessage.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            messages: Array.isArray(object === null || object === void 0 ? void 0 : object.messages)
                ? object.messages.map((e) => WakuMessage.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.messages) {
            obj.messages = message.messages.map((e) => e ? WakuMessage.toJSON(e) : undefined);
        }
        else {
            obj.messages = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseMessagePush();
        message.messages =
            ((_a = object.messages) === null || _a === void 0 ? void 0 : _a.map((e) => WakuMessage.fromPartial(e))) || [];
        return message;
    },
};
function createBaseFilterRPC() {
    return { requestId: "", request: undefined, push: undefined };
}
export const FilterRPC = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.requestId !== "") {
            writer.uint32(10).string(message.requestId);
        }
        if (message.request !== undefined) {
            FilterRequest.encode(message.request, writer.uint32(18).fork()).ldelim();
        }
        if (message.push !== undefined) {
            MessagePush.encode(message.push, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFilterRPC();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.requestId = reader.string();
                    break;
                case 2:
                    message.request = FilterRequest.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.push = MessagePush.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            requestId: isSet(object.requestId) ? String(object.requestId) : "",
            request: isSet(object.request)
                ? FilterRequest.fromJSON(object.request)
                : undefined,
            push: isSet(object.push) ? MessagePush.fromJSON(object.push) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.requestId !== undefined && (obj.requestId = message.requestId);
        message.request !== undefined &&
            (obj.request = message.request
                ? FilterRequest.toJSON(message.request)
                : undefined);
        message.push !== undefined &&
            (obj.push = message.push ? MessagePush.toJSON(message.push) : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseFilterRPC();
        message.requestId = (_a = object.requestId) !== null && _a !== void 0 ? _a : "";
        message.request =
            object.request !== undefined && object.request !== null
                ? FilterRequest.fromPartial(object.request)
                : undefined;
        message.push =
            object.push !== undefined && object.push !== null
                ? MessagePush.fromPartial(object.push)
                : undefined;
        return message;
    },
};
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
