// js-waku/build/esm/proto/waku/v2/light_push.js
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";
import { WakuMessage } from "../../waku/v2/message";
export const protobufPackage = "waku.v2";
function createBasePushRequest() {
    return { pubSubTopic: "", message: undefined };
}
export const PushRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.pubSubTopic !== "") {
            writer.uint32(10).string(message.pubSubTopic);
        }
        if (message.message !== undefined) {
            WakuMessage.encode(message.message, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePushRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pubSubTopic = reader.string();
                    break;
                case 2:
                    message.message = WakuMessage.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            pubSubTopic: isSet(object.pubSubTopic) ? String(object.pubSubTopic) : "",
            message: isSet(object.message)
                ? WakuMessage.fromJSON(object.message)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.pubSubTopic !== undefined &&
            (obj.pubSubTopic = message.pubSubTopic);
        message.message !== undefined &&
            (obj.message = message.message
                ? WakuMessage.toJSON(message.message)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBasePushRequest();
        message.pubSubTopic = (_a = object.pubSubTopic) !== null && _a !== void 0 ? _a : "";
        message.message =
            object.message !== undefined && object.message !== null
                ? WakuMessage.fromPartial(object.message)
                : undefined;
        return message;
    },
};
function createBasePushResponse() {
    return { isSuccess: false, info: "" };
}
export const PushResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.isSuccess === true) {
            writer.uint32(8).bool(message.isSuccess);
        }
        if (message.info !== "") {
            writer.uint32(18).string(message.info);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePushResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.isSuccess = reader.bool();
                    break;
                case 2:
                    message.info = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            isSuccess: isSet(object.isSuccess) ? Boolean(object.isSuccess) : false,
            info: isSet(object.info) ? String(object.info) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.isSuccess !== undefined && (obj.isSuccess = message.isSuccess);
        message.info !== undefined && (obj.info = message.info);
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBasePushResponse();
        message.isSuccess = (_a = object.isSuccess) !== null && _a !== void 0 ? _a : false;
        message.info = (_b = object.info) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBasePushRPC() {
    return { requestId: "", request: undefined, response: undefined };
}
export const PushRPC = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.requestId !== "") {
            writer.uint32(10).string(message.requestId);
        }
        if (message.request !== undefined) {
            PushRequest.encode(message.request, writer.uint32(18).fork()).ldelim();
        }
        if (message.response !== undefined) {
            PushResponse.encode(message.response, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePushRPC();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.requestId = reader.string();
                    break;
                case 2:
                    message.request = PushRequest.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.response = PushResponse.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            requestId: isSet(object.requestId) ? String(object.requestId) : "",
            request: isSet(object.request)
                ? PushRequest.fromJSON(object.request)
                : undefined,
            response: isSet(object.response)
                ? PushResponse.fromJSON(object.response)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.requestId !== undefined && (obj.requestId = message.requestId);
        message.request !== undefined &&
            (obj.request = message.request
                ? PushRequest.toJSON(message.request)
                : undefined);
        message.response !== undefined &&
            (obj.response = message.response
                ? PushResponse.toJSON(message.response)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBasePushRPC();
        message.requestId = (_a = object.requestId) !== null && _a !== void 0 ? _a : "";
        message.request =
            object.request !== undefined && object.request !== null
                ? PushRequest.fromPartial(object.request)
                : undefined;
        message.response =
            object.response !== undefined && object.response !== null
                ? PushResponse.fromPartial(object.response)
                : undefined;
        return message;
    },
};
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
