// js-waku/build/esm/proto/waku/v2/message.js
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";
export const protobufPackage = "waku.v2";
function createBaseWakuMessage() {
    return {
        payload: undefined,
        contentTopic: undefined,
        version: undefined,
        timestampDeprecated: undefined,
        timestamp: undefined,
    };
}
export const WakuMessage = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.payload !== undefined) {
            writer.uint32(10).bytes(message.payload);
        }
        if (message.contentTopic !== undefined) {
            writer.uint32(18).string(message.contentTopic);
        }
        if (message.version !== undefined) {
            writer.uint32(24).uint32(message.version);
        }
        if (message.timestampDeprecated !== undefined) {
            writer.uint32(33).double(message.timestampDeprecated);
        }
        if (message.timestamp !== undefined) {
            writer.uint32(80).sint64(message.timestamp);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWakuMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.payload = reader.bytes();
                    break;
                case 2:
                    message.contentTopic = reader.string();
                    break;
                case 3:
                    message.version = reader.uint32();
                    break;
                case 4:
                    message.timestampDeprecated = reader.double();
                    break;
                case 10:
                    message.timestamp = reader.sint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            payload: isSet(object.payload)
                ? bytesFromBase64(object.payload)
                : undefined,
            contentTopic: isSet(object.contentTopic)
                ? String(object.contentTopic)
                : undefined,
            version: isSet(object.version) ? Number(object.version) : undefined,
            timestampDeprecated: isSet(object.timestampDeprecated)
                ? Number(object.timestampDeprecated)
                : undefined,
            timestamp: isSet(object.timestamp)
                ? Long.fromString(object.timestamp)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.payload !== undefined &&
            (obj.payload =
                message.payload !== undefined
                    ? base64FromBytes(message.payload)
                    : undefined);
        message.contentTopic !== undefined &&
            (obj.contentTopic = message.contentTopic);
        message.version !== undefined &&
            (obj.version = Math.round(message.version));
        message.timestampDeprecated !== undefined &&
            (obj.timestampDeprecated = message.timestampDeprecated);
        message.timestamp !== undefined &&
            (obj.timestamp = (message.timestamp || undefined).toString());
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseWakuMessage();
        message.payload = (_a = object.payload) !== null && _a !== void 0 ? _a : undefined;
        message.contentTopic = (_b = object.contentTopic) !== null && _b !== void 0 ? _b : undefined;
        message.version = (_c = object.version) !== null && _c !== void 0 ? _c : undefined;
        message.timestampDeprecated = (_d = object.timestampDeprecated) !== null && _d !== void 0 ? _d : undefined;
        message.timestamp =
            object.timestamp !== undefined && object.timestamp !== null
                ? Long.fromValue(object.timestamp)
                : undefined;
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
const atob = globalThis.atob ||
    ((b64) => globalThis.Buffer.from(b64, "base64").toString("binary"));
function bytesFromBase64(b64) {
    const bin = atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
const btoa = globalThis.btoa ||
    ((bin) => globalThis.Buffer.from(bin, "binary").toString("base64"));
function base64FromBytes(arr) {
    const bin = [];
    for (const byte of arr) {
        bin.push(String.fromCharCode(byte));
    }
    return btoa(bin.join(""));
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
