// js-waku/build/main/lib/waku_message/symmetric.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateIv = exports.decrypt = exports.encrypt = exports.TagSize = exports.IvSize = exports.KeySize = void 0;
const crypto_1 = require("../crypto");
exports.KeySize = 32;
exports.IvSize = 12;
exports.TagSize = 16;
const Algorithm = { name: "AES-GCM", length: 128 };
async function encrypt(iv, key, clearText) {
    return (0, crypto_1.getSubtle)()
        .importKey("raw", key, Algorithm, false, ["encrypt"])
        .then((cryptoKey) => (0, crypto_1.getSubtle)().encrypt({ iv, ...Algorithm }, cryptoKey, clearText))
        .then((cipher) => new Uint8Array(cipher));
}
exports.encrypt = encrypt;
async function decrypt(iv, key, cipherText) {
    return (0, crypto_1.getSubtle)()
        .importKey("raw", key, Algorithm, false, ["decrypt"])
        .then((cryptoKey) => (0, crypto_1.getSubtle)().decrypt({ iv, ...Algorithm }, cryptoKey, cipherText))
        .then((clear) => new Uint8Array(clear));
}
exports.decrypt = decrypt;
function generateIv() {
    return (0, crypto_1.randomBytes)(exports.IvSize);
}
exports.generateIv = generateIv;
