// @xmtp/xmtp-js/dist/cjs/test/helpers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.newDevClient = exports.newLocalHostClient = exports.newLocalDockerClient = exports.CodecRegistry = exports.newWallet = exports.dumpStream = exports.waitForUserContact = exports.pollFor = exports.sleep = void 0;
const ethers_1 = require("ethers");
const src_1 = require("../src");
const utils_1 = require("../src/utils");
const assert_1 = __importDefault(require("assert"));
const LOCAL_DOCKER_MULTIADDR = '/ip4/127.0.0.1/tcp/9001/ws/p2p/16Uiu2HAmNCxLZCkXNbpVPBpSSnHj9iq4HZQj7fxRzw2kj1kKSHHA';
const LOCAL_HOST_MULTIADDR = '/ip4/127.0.0.1/tcp/9002/ws/p2p/16Uiu2HAmNCxLZCkXNbpVPBpSSnHj9iq4HZQj7fxRzw2kj1kKSHHA';
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
exports.sleep = sleep;
function pollFor(callback, timeoutMs, delayMs) {
    return __awaiter(this, void 0, void 0, function* () {
        const started = Date.now();
        try {
            return yield callback();
        }
        catch (err) {
            if (delayMs) {
                yield (0, exports.sleep)(delayMs);
            }
            const elapsedMs = Date.now() - started;
            const remainingTimeoutMs = timeoutMs - elapsedMs;
            if (remainingTimeoutMs <= 0) {
                throw new Error('timeout exceeded');
            }
            return yield pollFor(callback, remainingTimeoutMs, delayMs);
        }
    });
}
exports.pollFor = pollFor;
function waitForUserContact(c1, c2) {
    return __awaiter(this, void 0, void 0, function* () {
        return pollFor(() => __awaiter(this, void 0, void 0, function* () {
            const contact = yield c1.getUserContactFromNetwork(c2.address);
            assert_1.default.ok(contact);
            return contact;
        }), 20000, 200);
    });
}
exports.waitForUserContact = waitForUserContact;
function dumpStream(stream, timeoutMs = 1000) {
    return __awaiter(this, void 0, void 0, function* () {
        const messages = [];
        try {
            while (true) {
                const result = yield (0, utils_1.promiseWithTimeout)(timeoutMs, () => stream.next(), 'timeout');
                if (result.done) {
                    break;
                }
                messages.push(result.value);
            }
        }
        catch (_a) {
        }
        finally {
            stream.return();
        }
        return messages;
    });
}
exports.dumpStream = dumpStream;
function newWallet() {
    const key = src_1.PrivateKey.generate();
    if (!key.secp256k1) {
        throw new Error('invalid key');
    }
    return new ethers_1.Wallet(key.secp256k1.bytes);
}
exports.newWallet = newWallet;
// A helper to replace a full Client in testing custom content types,
// extracting just the codec registry aspect of the client.
class CodecRegistry {
    constructor() {
        this._codecs = new Map();
        this.registerCodec(new src_1.TextCodec());
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    registerCodec(codec) {
        const id = codec.contentType;
        const key = `${id.authorityId}/${id.typeId}`;
        this._codecs.set(key, codec);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    codecFor(contentType) {
        const key = `${contentType.authorityId}/${contentType.typeId}`;
        return this._codecs.get(key);
    }
}
exports.CodecRegistry = CodecRegistry;
// client running against local docker node,
// see dev/docker-compose
const newLocalDockerClient = () => src_1.Client.create(newWallet(), {
    bootstrapAddrs: [LOCAL_DOCKER_MULTIADDR],
});
exports.newLocalDockerClient = newLocalDockerClient;
// client running against local node running on the host,
// see github.com/xmtp/xmtp-node-go/scripts/xmtp-js.sh
const newLocalHostClient = () => src_1.Client.create(newWallet(), {
    bootstrapAddrs: [LOCAL_HOST_MULTIADDR],
});
exports.newLocalHostClient = newLocalHostClient;
// client running against the dev cluster in AWS
const newDevClient = () => src_1.Client.create(newWallet(), { env: 'dev' });
exports.newDevClient = newDevClient;
